### ISSUE TITLE
t.p.amp.AmpBox.serialize will _not_ return a string if any value is unicode

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3931]
|<img alt="dotz's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| dotz reported|
|-|-|
|Trac ID|trac#3931|
|Type|defect|
|Created|2009-07-25 22:16:16Z|
|Branch|https://github.com/twisted/twisted/tree/amp-unicode-3931|

If you place an unicode value in AmpBox, AmpBox.serialize will return an unicode string, not str. 

For example:

```
>>> from twisted.protocols.amp import AmpBox
>>> a = AmpBox(asdf = 'foo')
>>> type(a.serialize())
<type 'str'>
>>> a = AmpBox(asdf = u'foo')
>>> type(a.serialize())
<type 'unicode'>
```

This will cause funny errors on some platforms (see [#3930](https://github.com/twisted/twisted/issues/3930)).

This is also against AmpBox.serialize docstring, which says that it should return a str type, not unicode.
```
    def serialize(self):
        """
        Convert me into a wire-encoded string.

        @return: a str encoded according to the rules described in the module
        docstring.
        """
```

Please let me know what is the preferred way to get this fixed in trunk and I'll prepare a patch.

Attachments:

* [fix-3931.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/8d4/8d4709a9a9cfe773fd0b5d3982f448ade4e4eded/823cf1bb4920f39c41e902ce3394dfc6a5910f92.patch) (1526 bytes) - added by dotz on 2009-11-07 21:53:50Z - Let's check keys and values during serialize and if the strict mode is enabled, let's fail. Also includes an unit test
* [better-3931.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/8d4/8d4709a9a9cfe773fd0b5d3982f448ade4e4eded/31e1890deff8300c074f125851b420a1a02640f6.patch) (1457 bytes) - added by MostAwesomeDude on 2011-02-11 16:49:37Z - Improved patch, with style and other requested changes
* [fixed-3931.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/8d4/8d4709a9a9cfe773fd0b5d3982f448ade4e4eded/7fdecc781a82c1eed059dce8baf017a080245d97.patch) (2004 bytes) - added by MostAwesomeDude on 2011-03-15 21:52:14Z - Patch against 3931 branch with fixes from review comments

<details><summary>Searchable metadata</summary>

```
trac-id__3931 3931
type__defect defect
reporter__dotz dotz
priority__normal normal
milestone__ 
branch__branches_amp_unicode_3931 branches/amp-unicode-3931
branch_author__dotz__MostAwesomeDude dotz, MostAwesomeDude
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1248560176000000 1248560176000000
changetime__1300253348000000 1300253348000000
version__None None
owner__lvh lvh
cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="dotz's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|dotz commented|
|-|-|

Do you need anything else from me to get this one small change merged?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph set owner to dotz|
|-|-|

Best solution: don't put unicode strings into AmpBox :).

Thanks for poking this issue, I certainly would have forgotten about it.  I am not going to fix it myself.  In most Python libraries, these sorts of checks are left out; if you put other random 'str' subclasses or non-string objects into an `AmpBox` you're going to get bogus results too, this isn't unique to '`unicode`' objects.

If this is a problem that you're running into frequently, then please feel free to add a "strict" or "debug" mode which will reject anything but str:str mappings in an AmpBox; make sure that you cover `.update`, `.__setitem__`, `.setdefault`, and any other methods that dicts might have that I've forgotten about.  Submit a patch, with tests, submit it for review.

**Comment by twisted-trac** [comment]
|<img alt="dotz's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|dotz removed owner|
|-|-|

I added a patch, that makes debugging a lot easier. Thanks for your time!

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve set owner to @glyph|
|-|-|

I don't feel super great about this "strict" mode, so I'll let glyph review this.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to dotz|
|-|-|

I think the "strict" mode is silly too, particularly if it defaults to on.  Just do the checks all the time.  Doing it in `serialize` is a bit later than would be ideal, but it has the advantage of avoiding the need to do it in every single `dict` mutation method.  So I guess I like it.  Please use the one-argument form of `raise`, though, and include the name of the offending key.

**Comment by twisted-trac** [comment]
|[<img alt="MostAwesomeDude's avatar" src="https://avatars.githubusercontent.com/u/118035?s=50" width="50" height="50">](https://github.com/MostAwesomeDude)<a name="note_6"></a>|@MostAwesomeDude commented|
|-|-|

Replying to [exarkun](#note_5):
> I think the "strict" mode is silly too, particularly if it defaults to on.  Just do the checks all the time.  Doing it in `serialize` is a bit later than would be ideal, but it has the advantage of avoiding the need to do it in every single `dict` mutation method.  So I guess I like it.  Please use the one-argument form of `raise`, though, and include the name of the offending key.

New patch attached; review please. This possibly is biting Bravo users on Win32, and I feel like fixing upstream bugs today. :3

**Comment by twisted-trac** [comment]
|[<img alt="MostAwesomeDude's avatar" src="https://avatars.githubusercontent.com/u/118035?s=50" width="50" height="50">](https://github.com/MostAwesomeDude)<a name="note_7"></a>|@MostAwesomeDude commented|
|-|-|

I am bad with Trac. :c

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [30701]) Branching to 'amp-unicode-3931'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [30702]) Apply better-3931.patch

refs [#3931](https://github.com/twisted/twisted/issues/3931)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set owner to @MostAwesomeDude|
|-|-|

  1. Can you give `AmpBoxTest` a class docstring (and maybe pluralize the name, since it has more than one test method ;)
  1. There's a test for the value being unicode, but not for the key.  Can you add another test for the key case?
  1. I think it would probably be safer to use %r for the value, since the TypeError could end up going pretty much anywhere, and having non-ascii unicode in it probably isn't safe in many cases (consider the behavior of `str(TypeError(u'\N{SNOWMAN}'))`).  It'd be nice to have a test for this possible error case too.
  1. Unless you particularly disagree, I think putting the key into the error message in both cases is better.  Even knowing the value of the value may not make it much easier to track down, whereas knowing that the corresponding key is should help a lot.

Thanks very much for picking this up!  Since there's a sprint tomorrow, it's quite likely that if it's ready for another review by then someone will take another look at it quickly. :)

**Comment by twisted-trac** [comment]
|[<img alt="MostAwesomeDude's avatar" src="https://avatars.githubusercontent.com/u/118035?s=50" width="50" height="50">](https://github.com/MostAwesomeDude)<a name="note_11"></a>|@MostAwesomeDude commented|
|-|-|

Replying to [exarkun](#note_10):
>   1. Can you give `AmpBoxTest` a class docstring (and maybe pluralize the name, since it has more than one test method ;)
>   1. There's a test for the value being unicode, but not for the key.  Can you add another test for the key case?
>   1. I think it would probably be safer to use %r for the value, since the TypeError could end up going pretty much anywhere, and having non-ascii unicode in it probably isn't safe in many cases (consider the behavior of `str(TypeError(u'\N{SNOWMAN}'))`).  It'd be nice to have a test for this possible error case too.
>   1. Unless you particularly disagree, I think putting the key into the error message in both cases is better.  Even knowing the value of the value may not make it much easier to track down, whereas knowing that the corresponding key is should help a lot.
> 
> Thanks very much for picking this up!  Since there's a sprint tomorrow, it's quite likely that if it's ready for another review by then someone will take another look at it quickly. :)

Sorry for taking so long on this. :c

 1. Done.
 1. Done.
 1. Agreed; I've made that change. I'm not sure how to test that the raised exception is ASCII-safe though.
 1. Done; I'm not sure I like the lack of symmetry, but I think the messages are definitely clear enough at this point.

Review please?

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_12"></a>|@lvh set owner to @lvh<br>@lvh set status to `assigned`|
|-|-|

1.Yay!
 2. Yay!
 3. I agree it is. Yay!
 4. I agree. Yay!

I will merge this.

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_13"></a>|@lvh set status to `closed`|
|-|-|

(In [31176]) 
Complain consistently when t.p.amp.AmpBox.serialize gets fed Unicode.

Author: MostAwesomeDude
Reviewer: lvh, exarkun
Fixes: [#3931](https://github.com/twisted/twisted/issues/3931)
